### ISSUE TITLE
[Bug] PDF download does not work locally with docker compose

### ISSUE THREAD

**Issue by esep** [AmruthPillai/Reactive-Resume#1431]
### Is there an existing issue for this?

- [X] Yes, I have searched the existing issues

### Product Variant

Self-Hosted

### Current Behavior

Cannot download resume as PDF in a docker compose setup: "Failed to load resource: the server responded with a status of 504"

### Expected Behavior

Resume is downloaded as a PDF file

### Steps To Reproduce

1. Clone repo
2. Start with docker compose up
3. Open browser (e.g. Chrome) to localhost:3000
4. Sign up a test user
5. Create a resume with any data
6. Download PDF
7. Get the following error in the browser developer console and red cross popup on the lower right without text `:3100/printer/testi/test?lastUpdated=1692103045:1     Failed to load resource: the server responded with a status of 504 (Gateway Timeout)`
8. Downloading does not start

### What browsers are you seeing the problem on?

Chrome

### Anything else?

![RxResume pdf error](https://github.com/AmruthPillai/Reactive-Resume/assets/15034165/b645bb89-cb99-412e-9bc9-632ba46af218)

**Comment by NateTheSage** [comment]
Same here, I just had to reinstall from scratch recently and it fails almost immediately with 504.


![chrome_MMEY3CW8DZ](https://github.com/AmruthPillai/Reactive-Resume/assets/42617513/eb5a3346-67b8-460f-8c0f-1c52e608ed83)

I think I had it working a while ago, but I can't recall.

I'm using a docker-compose file which includes Traefik:
```
version: "3.8"

services:
  postgres:
    image: postgres:alpine
    container_name: rxresume-db
    restart: always
    expose:
      - 5432
    volumes:
      - ${DOCKER_VOLUME_STORAGE:-/mnt/docker-volumes}/rxresume/postgresql:/var/lib/postgresql/data
#      - pgdata:/var/lib/postgresql/data
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      start_period: 15s
      interval: 30s
      timeout: 30s
      retries: 3
    environment:
      - POSTGRES_DB=postgres
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
    networks:
      - proxy

  server:
    image: amruthpillai/reactive-resume:server-latest
    container_name: rxresume-server
    restart: always
    extra_hosts:
      - "resume.example.net:<snip>" # optionally enable if PDF export keeps failing; specify your domain and server's IP address where traefik is running
    expose:
      - 3100
    depends_on:
      - postgres
    environment:
      - PUBLIC_URL=https://resume.example.net
      - PUBLIC_SERVER_URL=https://resume.example.net/api # only change the subdomain, leave /api as is
      - PUBLIC_GOOGLE_CLIENT_ID=
      - POSTGRES_DB=postgres
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
      - SECRET_KEY=<snip>
      - POSTGRES_HOST=postgres
      - POSTGRES_PORT=5432
      - POSTGRES_SSL_CERT=
      - JWT_SECRET=<snip>
      - JWT_EXPIRY_TIME=604800
      - GOOGLE_CLIENT_SECRET=
      - GOOGLE_API_KEY=
      - MAIL_FROM_NAME=Reactive Resume
      - MAIL_FROM_EMAIL=noreply@rxresu.me
      - MAIL_HOST=
      - MAIL_PORT=
      - MAIL_USERNAME=
      - MAIL_PASSWORD=
      - STORAGE_BUCKET=
      - STORAGE_REGION=
      - STORAGE_ENDPOINT=
      - STORAGE_URL_PREFIX=
      - STORAGE_ACCESS_KEY=
      - STORAGE_SECRET_KEY=
      - PDF_DELETION_TIME=
    networks:
      - proxy
    labels:
      - traefik.enable=true
      - traefik.http.routers.rxresume-server.rule=Host(`resume.example.net`) && PathPrefix(`/api`) # only change the subdomain, leave /api as is
      - traefik.http.services.rxresume-server.loadbalancer.server.port=3100
      - traefik.docker.network=proxy
      # Part for optional traefik middlewares
      - traefik.http.routers.rxresume-server.middlewares=path-strip # may add local-ipwhitelist@file for access control
      - traefik.http.middlewares.path-strip.stripprefix.prefixes=/api
      - traefik.http.middlewares.path-strip.stripprefix.forceSlash=false

  client:
    image: amruthpillai/reactive-resume:client-latest
    container_name: rxresume-client
    restart: always
    extra_hosts:
      - "resume.example.net:<snip>" # # optionally enable if PDF export keeps failing; specify your domain and server's IP address where traefik is running
    expose:
      - 3000
    depends_on:
      - server
    environment:
      - PUBLIC_URL=https://resume.example.net
      - PUBLIC_SERVER_URL=https://resume.example.net/api # only change the subdomain, leave /api as is
      - PUBLIC_GOOGLE_CLIENT_ID=
    networks:
      - proxy
    labels:
      - traefik.enable=true
      - traefik.http.routers.rxresume-client.rule=Host(`resume.example.net`)
      - traefik.http.services.rxresume-client.loadbalancer.server.port=3000
      - traefik.docker.network=proxy
      # Part for optional traefik middlewares
      #- traefik.http.routers.rxresume-client.middlewares=local-ipwhitelist@file # may enable this middleware for access control

  traefik:
    image: traefik:v2.10.1
#    network_mode: "host"
    container_name: rxresume-traefik
    restart: unless-stopped
    command:
      - "--log.level=INFO"
      - "--api.insecure=true"
      - "--providers.docker=true"
      - "--providers.docker.exposedbydefault=false"
      - "--entrypoints.web.address=:80"
    ports:
#      - 127.0.0.1:8080:80
      - 8080:80
#      - 127.0.0.1:8080:8080
    environment:
      - VIRTUAL_HOST=resume.example.net
      - VIRTUAL_PORT=80
    networks:
      - proxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro

  watchtower:
    image: containrrr/watchtower
    container_name: watchtower
    restart: always
    environment:
      TZ: America/Denver
      WATCHTOWER_CLEANUP: "true"
      WATCHTOWER_DEBUG: "false"
      WATCHTOWER_NOTIFICATIONS: "email"
      WATCHTOWER_NOTIFICATION_EMAIL_FROM: "watchtower@example.net"
      WATCHTOWER_NOTIFICATION_EMAIL_TO: "sysadmin@example.net"
      # you have to use a network alias here, if you use your own certificate 
      WATCHTOWER_NOTIFICATION_EMAIL_SERVER: "mail.example.net"
      WATCHTOWER_NOTIFICATION_EMAIL_SERVER_PORT: "25"
      WATCHTOWER_NOTIFICATION_EMAIL_DELAY: 2
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock

#volumes:
#  pgdata:

networks:
  proxy:
    external: true
```

The containers show no errors of any kind at present.

**Comment by wzzrd** [comment]
I tried rolling back to earlier versions today, to see if that would help, but it doesn't. I had this working, too. Please let us know what you need in order to debug this. As stated above: container logs show nothing when the request is made, and the response reads:

```json
{
  "statusCode":504,
  "timestamp":"2023-09-01T10:16:51.108Z",
  "path":"/printer/something/public-resume-v2?lastUpdated=1679045438"
}
```

**Comment by michael-robbins** [comment]
Getting this exact same error, nothing in client or server logs, just a HTTP 504 from Express from the client container

**Comment by xorgxx** [comment]
so what is the next step ?

**Comment by esturniolo** [comment]
Same here.
Using the compose example without changing anything.

**Comment by aalmazan** [comment]
This is happening because the server will try to create a PDF by visiting the local site and rendering it as pages.

You can fix that initial issue by changing the `PUBLIC_URL` in `compose.yml`: https://github.com/AmruthPillai/Reactive-Resume/blob/main/compose.yml#L16-L25

Change `http://localhost:3000` -> `http://client:3000`


After that, unfortunately it's a tougher to debug and I haven't been able to get meaningful errors out. It seems to error out on the lines below:

https://github.com/AmruthPillai/Reactive-Resume/blob/main/server/src/printer/printer.service.ts#L100-L101

**Comment by mike59881051** [comment]
Encounter simliar issue the other day, log off and restart the container fixed temporary

**Comment by otterDeveloper** [comment]
I think I got it.
TLDR: Beyond what @aalmazan described, this also occurs due to the use of special characters in your `SECRET_KEY` environment variable while having the 'Is Publicly Accessible?' option turned off.

I console logged https://github.com/AmruthPillai/Reactive-Resume/blob/main/server/src/printer/printer.service.ts#L100
which builds the following url:

` http://localhost:3000/otter/test1/printer?secretKey=lfViZcyaF25btZcM3+EcfqV/N5zkvvH0N28emnVHcPs=`

The issue arises because the secret key is not being URL encoded.

I generated my secret key using `openssl rand -base64 32`, which appends an `=` at the end. 

Making your resume public appears to skip the secret key verification.

Quick Fixes:
* Make your resume public.
* Alter your secret key to exclude special characters.

I will open a pr to fix this.

Finger crossed there aren't further complications 🤞🏻.

**Comment by NateTheSage** [comment]
That is absolutely wild on both accounts. I'm going to try this later tonight and see if this resolves it. It should have probably occurred to me using Docker that the address for the `PUBLIC_URL` should be the container.

Will report back with results.

**Comment by NateTheSage** [comment]
> This is happening because the server will try to create a PDF by visiting the local site and rendering it as pages.
> 
> You can fix that initial issue by changing the `PUBLIC_URL` in `compose.yml`: https://github.com/AmruthPillai/Reactive-Resume/blob/main/compose.yml#L16-L25
> 
> Change `http://localhost:3000` -> `http://client:3000`
> 
> After that, unfortunately it's a tougher to debug and I haven't been able to get meaningful errors out. It seems to error out on the lines below:
> 
> https://github.com/AmruthPillai/Reactive-Resume/blob/main/server/src/printer/printer.service.ts#L100-L101

This fixed it for me, I was surprised to see the prompt for saving the PDF to my local disk!

It should be noted that for other Docker users that you only need to change `PUBLIC_URL`, your `PUBLIC_SERVER_URL` entry under the `server` container should stay the same, especially if you're doing SSL termination via reverse proxy, otherwise you're going to get mixed content warnings.

As for the second option, I couldn't test this as my key already has no special characters in it.

**Comment by 0xERR0R** [comment]
As workaround, I changed variables "PUBLIC_URL" and "PUBLIC_SERVER_URL" from "localhost" to my host IP address and it worked for me

**Comment by cacpmw** [comment]
> I think I got it. TLDR: Beyond what @aalmazan described, this also occurs due to the use of special characters in your `SECRET_KEY` environment variable while having the 'Is Publicly Accessible?' option turned off.
> 
> I console logged https://github.com/AmruthPillai/Reactive-Resume/blob/main/server/src/printer/printer.service.ts#L100 which builds the following url:
> 
> ` http://localhost:3000/otter/test1/printer?secretKey=lfViZcyaF25btZcM3+EcfqV/N5zkvvH0N28emnVHcPs=`
> 
> The issue arises because the secret key is not being URL encoded.
> 
> I generated my secret key using `openssl rand -base64 32`, which appends an `=` at the end.
> 
> Making your resume public appears to skip the secret key verification.
> 
> Quick Fixes:
> 
>     * Make your resume public.
> 
>     * Alter your secret key to exclude special characters.
> 
> 
> I will open a pr to fix this.
> 
> Finger crossed there aren't further complications 🤞🏻.

Making it as public fixed it for me. I changed secrets to not include special chars but that alone didnt take effect.

**Comment by AmruthPillai** [comment]
This issue was automatically closed as it was created before 2024-01-01.
